### PR TITLE
GIX-1219: Manage errors when fetching SNS proposal data

### PR DESCRIPTION
**Pull request by lmuntaner**
# Motivation

User should be redirected to the SNS proposals list if there is some problem loading the specific proposal.

# Changes

Changes in SnsProposalDetail.svelte page:
* Redirect if the parameter can't be parsed to a bigint.
* Redirect if there is an error when loading the proposal

# Tests

* Add test for SnsProposalDetail.svelte page. Add test case for each of the above scenarios.
* Add error capabilities to the sns governance api fake.

### CODE REVIEW

**Comment by lmuntaner** [comment]
@dskloetd please review

File: frontend/src/lib/pages/SnsProposalDetail.svelte [hunk 1]
```diff
@@ -23,9 +24,21 @@
     }
   });
 
+  // By setting a local variable, we avoid calling the below text when the whole store is changes but the value is the same.
```
Comment (dskloetd) [inline frontend/src/lib/pages/SnsProposalDetail.svelte:27]: This comment is very unclear. Can you be specific?
What changes?
What stays the same?
What "text" (?) do we need to avoid calling and why?

File: frontend/src/lib/pages/SnsProposalDetail.svelte [hunk 2]
```diff
@@ -23,9 +24,21 @@
...
+  let rootCanisterIdText: undefined | string;
```
Comment (dskloetd) [inline frontend/src/lib/pages/SnsProposalDetail.svelte:28]: Could we instead have
```
$: let rootCanisterIdText = $snsOnlyProjectStore.toText();
```

And then just trigger getSnsProposalById when $rootCanisterIdText changes (and, I guess, reconstruct the Principal from the text)? Because it's a string I think it won't have the issue of the Principal that it can be equal but different.

File: frontend/src/lib/pages/SnsProposalDetail.svelte [hunk 3]
```diff
@@ -23,9 +24,21 @@
...
-    if (nonNullish(proposalIdText) && nonNullish($snsOnlyProjectStore)) {
+    if (
+      nonNullish(proposalIdText) &&
+      nonNullish($snsOnlyProjectStore) &&
+      // Call the block only if the root canister id has changed.
+      $snsOnlyProjectStore.toText() !== rootCanisterIdText
+    ) {
+      rootCanisterIdText = $snsOnlyProjectStore.toText();
+      // We need this to be used in the handleError callback.
+      // Otherwise, TS doesn't believe that the value of `rootCanisterIdText` won't change.
```
Comment (dskloetd) [inline frontend/src/lib/pages/SnsProposalDetail.svelte:40]: And it's right about that, right? It actually might change.
Shall we call it something like rootCanisterIdAtTimeOfRequest?

File: frontend/src/lib/pages/SnsProposalDetail.svelte [hunk 4]
```diff
@@ -23,9 +24,21 @@
...
+      const scopedId = rootCanisterIdText;
```
Comment (dskloetd) [inline frontend/src/lib/pages/SnsProposalDetail.svelte:41]: Optional: Instead you could also create the `{universe: rootCanisterIdText}` object here and use it both in line 52 and line 65.

File: frontend/src/tests/fakes/sns-governance-api.fake.ts [hunk 5]
```diff
@@ -235,6 +242,31 @@ async function queryProposals({
...
+  rootCanisterId,
+  certified: _,
+  proposalId,
+}: {
+  rootCanisterId: Principal;
+  identity: Identity;
+  certified: boolean;
+  proposalId: SnsProposalId;
+}): Promise<SnsProposalData | undefined> {
+  if (errorsMap.has(ImplementationsKeys.queryProposal)) {
```
Comment (dskloetd) [inline frontend/src/tests/fakes/sns-governance-api.fake.ts:256]: How about still using mockRejectOnce etc. for errors?
You can still define mock behavior the old way as well.
And errors you often don't want to persist so using the "once" functions of jest is convenient.
And I expect after 1 error happens, it will automatically go back to the fake provided behavior.
And you can of course always also just use a non-existed proposal ID to trigger an error.

File: frontend/src/tests/lib/pages/SnsProposalDetail.spec.ts [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+import * as snsGovernanceFake from "$tests/fakes/sns-governance-api.fake";
+import { mockAuthStoreNoIdentitySubscribe } from "$tests/mocks/auth.store.mock";
+import { mockCanisterId } from "$tests/mocks/canisters.mock";
+import { blockAllCallsTo } from "$tests/utils/module.test-utils";
+import { render, waitFor } from "@testing-library/svelte";
+import { get } from "svelte/store";
+
+jest.mock("$lib/api/sns-governance.api");
+
+const blockedApiPaths = ["$lib/api/sns-governance.api"];
```
Comment (dskloetd) [inline frontend/src/tests/lib/pages/SnsProposalDetail.spec.ts:19]: You don't need to block this if you also install the same fake.

File: frontend/src/tests/lib/pages/SnsProposalDetail.spec.ts [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+          proposalIdText: "invalid",
+        },
+      });
+      await waitFor(() => {
+        const { path } = get(pageStore);
+        return expect(path).toEqual(AppPath.Proposals);
+      });
+    });
+
+    it("should redirect to the list of sns proposals if proposal is not found", async () => {
```
Comment (dskloetd) [inline frontend/src/tests/lib/pages/SnsProposalDetail.spec.ts:48]: Can you also add a test that does not have an error?
Because as currently tested it would also pass if there is always a redirect.

File: frontend/src/lib/pages/SnsProposalDetail.svelte (hunk 2 above)
Comment (lmuntaner) [reply frontend/src/lib/pages/SnsProposalDetail.svelte:28]: I wanted to avoit reconstructing the Principal, but sure, why not.

File: frontend/src/lib/pages/SnsProposalDetail.svelte (hunk 1 above)
Comment (lmuntaner) [reply frontend/src/lib/pages/SnsProposalDetail.svelte:27]: sure, let me try again

File: frontend/src/lib/pages/SnsProposalDetail.svelte (hunk 3 above)
Comment (lmuntaner) [reply frontend/src/lib/pages/SnsProposalDetail.svelte:40]: Yes, it's correct. Sure, I can rename

File: frontend/src/tests/fakes/sns-governance-api.fake.ts (hunk 5 above)
Comment (lmuntaner) [reply frontend/src/tests/fakes/sns-governance-api.fake.ts:256]: I can throw an error if not found. I believe that's how the api behaves actually.

File: frontend/src/tests/lib/pages/SnsProposalDetail.spec.ts (hunk 7 above)
Comment (lmuntaner) [reply frontend/src/tests/lib/pages/SnsProposalDetail.spec.ts:48]: I will do that in the next PR with the skeleton, page object and placeholder for the content.

**Comment by lmuntaner** [comment]
@dskloetd ready for another review

**Review by dskloetd — APPROVED** [review]
